tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+        )]
+      \* Remove message from the queue.
+      /\ subnets' = [
+            subnets EXCEPT ![s].msgsToLedger = Tail(@)
+         ]
+      /\ UNCHANGED<<subnetMsgs>>
+
+SubnetReceiveApprove == 
+  /\ Len(subnetMsgs) > 0
+  /\ LET msg == Head(subnetMsgs) IN
+      msg.type = "approve"
+      /\ subnets' = [
+        subnets EXCEPT ![msg.to].unfinalized[msg.from] = @ - msg.amount,
+                       ![msg.to].balance = @ + msg.amount
+      ]
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = Tail(subnetMsgs)
+      /\ UNCHANGED<<ledger>>
+
+LedgerReceiveMessage ==
+  \* If a message is there and the balance checks out, then update it and
+  \* send approval to the subnet. If they don't check out, add sender to
+  \* the dishonestSubnets set and don't update balances.
+  LET
+    msg == Head(ledger.msgs)
+  IN
+   /\ Len(ledger.msgs) > 0
+   
+   /\ IF msg.amount > ledger.balances[msg.from] THEN
+        
+        
+        /\ ledger' = [ledger EXCEPT
+               \* Remove msg from queue.
+               !["msgs"] = Tail(@),
+               
+               \* Subnet spent more than it has. Mark it as a bad subnet.
+               !["dishonestSubnets"] = @ \union {msg.from}
+           ]
+        /\ UNCHANGED<<subnets, subnetMsgs>>
+      ELSE
+        /\ ledger' = [
+            ledger EXCEPT
+              \* Transaction is valid. Update ledger balances.
+              !["balances"][msg.from] = @ - msg.amount,
+              !["balances"][msg.to] = @ + msg.amount,
+              
+              \* Remove msg from queue.
+              !["msgs"] = Tail(@)
+            ]
+        /\ subnetMsgs'= Append(subnetMsgs,
+            [type |-> "approve", from |-> msg.from, to |-> msg.to, amount |-> msg.amount]
+           )
+        /\ UNCHANGED<<subnets>>
+
+Next ==
+  \/ SubnetSendTransfer
+  \/ SubnetReceiveTransfer
+  \/ SubnetReceiveApprove
+  \/ SubnetDishonestSendTransfer
+  \/ LedgerReceiveMessage
+  \/ LedgerPoll
+
+BalancesNonNegative ==
+  (*************************************************************************)
+  (* Invariant to ensure that all subnet balances are non-negative.        *)
+  (*************************************************************************)
+  \A s \in SUBNETS: subnets[s].balance >= 0
+
+
+LedgerIdentifiesAllDishonestSubnets ==
+  {s \in SUBNETS: ~subnets[s].honest} = ledger.dishonestSubnets
+
+LedgerIdentifiesAllDishonestSubnetsEventually ==
+  \* A temporal property ensuring that all dishonest subnets are,
+  \* eventually, identified by the ledger.
+  \*
+  \* NOTE: I think this invariant is not relevant anymore and can be removed.
+  <>[][LedgerIdentifiesAllDishonestSubnets]_<<subnets, subnetMsgs, ledger>>
+
+RECURSIVE RecordToSeq(_,_)
+RecordToSeq(f, ks) ==
+  \* Helper method to transform a record's values into a sequence.
+  IF ks = {} THEN << >>
+  ELSE << f[CHOOSE k \in ks: TRUE] >> \o RecordToSeq(f, ks \ {CHOOSE k \in ks: TRUE})
+
+RECURSIVE SumSeq(_)
```

Comment:
You could replace these next few definitions with a definition from a sorta-stdlib for TLA:
https://github.com/tlaplus/CommunityModules/blob/3fb30f52426b45cda44541b906ad9ebe16ce07c4/modules/FiniteSetsExt.tla#L40C1-L40C14
See also:
https://github.com/dfinity/formal-models/blob/master/tla/init-vscode.sh
to make this work out of the box (hopefully!) in VSCode.

Reply:
I want to do that, but the method you're referring to is for a set, whereas here I'm dealing with records and sequences. For some reason I didn't see similar definitions on sequences, and records don't seem to have any helper methods?

Reply:
I think you can just do `MapThenSumSet(LAMBDA s: ledger.balances[s], SUBNETS)`?

Reply:
Ah git blame tells me this has only been added 4 months ago:
https://github.com/tlaplus/CommunityModules/blame/master/modules/FiniteSetsExt.tla
so the JAR is outdated. I've bumped up the version in master now so it should work if you merge it.